### PR TITLE
Initial support of setting KeepIndentInBranch.

### PR DESCRIPTION
**Pull request by nojaf**
A first attempt to get to the behaviour of https://github.com/fsprojects/fantomas/issues/1361.
I would need some more examples to detect if all use-cases are covered by this.
The solution is based on the AST of the source so that is still in spirit with the project.
Please take a look at the unit tests and let me know if this is a good start.

### CODE REVIEW

File: src/Fantomas.Tests/IfThenElseTests.fs [hunk 1]
```diff
@@ -1651,14 +1651,14 @@ let tryDecompile (ty: FSharpEntity) =
     match ty.TryFullName with
     | Some fullName -> return decompile ty.Assembly.SimpleName externalSym
     | None ->
-        // might be abbreviated type (like string)
```
Comment (Smaug123) [inline src/Fantomas.Tests/IfThenElseTests.fs:1654]: Just checking that in the final version, you'll also keep the tests for when the feature is turned off? (I'm happy with this test case though.)

File: src/Fantomas.Tests/ComputationExpressionTests.fs [hunk 2]
```diff
@@ -1609,25 +1609,25 @@ let rec loop () =
 
     match msg with
     | Handle (eventSource, command, reply) ->
-        let! stream = eventSource |> eventStore.GetStream
```
Comment (Smaug123) [inline src/Fantomas.Tests/ComputationExpressionTests.fs:1612]: Looks sane to me

File: src/Fantomas.Tests/KeepIndentInBranch.fs [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        equal
+        """
+if someCondition then
+    0
+else
+let config =
+    Configuration.Read "/myfolder/myfile.xml"
+
+let result = Process.main config otherArg
+if result.IsOk then 0 else -1
```
Comment (Smaug123) [inline src/Fantomas.Tests/KeepIndentInBranch.fs:39]: I think some of the value of this test case has been lost with how short `0` and `-1` are. Maybe it's worth making these expressions long enough that Fantomas splits them onto new lines, so we can see how they behave?

File: src/Fantomas.Tests/KeepIndentInBranch.fs [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    let args = parse argv
+
+    let instructions = Library.foo args
+
+    if args.DryRun = RunMode.Dry then
+        printfn "Would execute actions, but --dry-run was supplied: %+A" instructions
+        0
+    else
+    // proceed with main method
+    let output = Library.execute instructions
```
Comment (Smaug123) [inline src/Fantomas.Tests/KeepIndentInBranch.fs:87]: Happy with this one

File: src/Fantomas.Tests/KeepIndentInBranch.fs [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+
+    let instructions = Library.foo args
+
+    match args.DryRun with
+    | RunMode.Dry ->
+        printfn "Would execute actions, but --dry-run was supplied: %+A" instructions
+        0
+    | RunMode.Wet ->
+    // proceed with main method
+    let output = Library.execute instructions
```
Comment (Smaug123) [inline src/Fantomas.Tests/KeepIndentInBranch.fs:139]: Happy with this one

File: src/Fantomas.Tests/KeepIndentInBranch.fs [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+type RunMode =
+    | Dry
+    | Wet
+
+let main argv =
+    let args = parse argv
+
+    let instructions = Library.foo args
+
+    match! args.DryRun with
```
Comment (Smaug123) [inline src/Fantomas.Tests/KeepIndentInBranch.fs:185]: Astonished this compiles, but sure

File: src/Fantomas.Tests/KeepIndentInBranch.fs [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+    let instructions = Library.foo args
+
+    let exitCode =
+        function
+        | RunMode.Dry ->
+            printfn "Would execute actions, but --dry-run was supplied: %+A" instructions
+            0
+        | RunMode.Wet ->
+        // proceed with main method
```
Comment (Smaug123) [inline src/Fantomas.Tests/KeepIndentInBranch.fs:247]: Seems reasonable

File: src/Fantomas.Tests/PatternMatchingTests.fs [hunk 8]
```diff
@@ -587,8 +587,8 @@ match x with
         """
 match x with
 | Some y ->
-    let z = 1
```
Comment (Smaug123) [inline src/Fantomas.Tests/PatternMatchingTests.fs:590]: I'm not convinced by this. I think the feature should only affect the last block, and it shouldn't affect the *width* of an indentation at all - only the "number of tabs". This is the first block in the match, and not the last block, so I think it shouldn't change.

**Review by Smaug123 — COMMENTED** [review]
The implementation is surprisingly simple!

I guess it would be useful to have a few more tests documenting the boundary conditions. For example, the difference (if any) between the following:

```fsharp
match true with
| false ->
    failwith ""
| true ->
    failwith ""

foo ()
```
and 
```fsharp
match true with
| false ->
    failwith ""
| true ->
    failwith ""
    0

foo ()
```

and also what happens when you remove the trailing `foo ()`. The test cases right now are more showing us what the feature does, which is very useful, but it's the edge cases where the subtlety lies. As I understand the feature, for example, my version with the extra line that returns `0` should trigger unindentation, but the version without should not; and the presence of `foo ()` must prevent unindentation. But I recognise that I have really not specified what the feature should do, so if my understanding is different from yours, that's probably not a problem!

File: src/Fantomas.Tests/ComputationExpressionTests.fs (hunk 2 above)
Comment (nojaf) [reply src/Fantomas.Tests/ComputationExpressionTests.fs:1612]: This is actually a bit of a side effect of a change in CodePrinter.
I saw the current behavior is somewhat weird and logged https://github.com/fsprojects/fantomas/issues/1501 as it is not in line with the style guides.
So, the modified tests are more a result of that change.

File: src/Fantomas.Tests/PatternMatchingTests.fs (hunk 8 above)
Comment (nojaf) [reply src/Fantomas.Tests/PatternMatchingTests.fs:590]: This test is actually unrelated to the feature but also is a side effect of solving https://github.com/fsprojects/fantomas/issues/1501.

File: src/Fantomas/CodePrinter.fs [hunk 9]
```diff
@@ -4373,12 +4378,32 @@ and genClause astContext hasBar (Clause (p, e, eo) as ce) =
         +> fun ctx -> leaveNodeTokenByName (arrowRange ctx) RARROW ctx
 
     genTriviaBeforeClausePipe p.Range
-    +> ifElse
-        hasBar
-        (sepBar +> atCurrentColumnWithPrepend pat body
```
Comment (nojaf) [inline src/Fantomas/CodePrinter.fs:4378]: I remove `atCurrentColumnWithPrepend` to solve https://github.com/fsprojects/fantomas/issues/1501 and because it simplifies the flow of the code. 
The only reason we had this was because one nitpick of a user in a time where style guides weren't a thing yet.
When I looked it up in the style I noticed our current behavior was inconsistent.

`AtCurrentColumn` will lock the column indent and override the normal flow of indentation, it is a last resort and that is the main reason I wanted to get rid of it.

**Comment by Smaug123** [comment]
One of our examples:

```fsharp
module Foo =
    let make () =
        { new IBar with
            member __.DoIt (i : int) =
                doSomeStuff ()
                match i with
                | 0 ->
                    handleIt ()
                    None
                | _ ->

                let baz = foo ()
                match int baz with
                | 0 ->
                    handleAgain ()
                    None
                | _ ->

                doEvenMore ()
                Some 3
        }
```
